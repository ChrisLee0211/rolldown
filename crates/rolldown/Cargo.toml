[package]
edition = "2021"
name    = "rolldown"
version = "0.1.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
derivative                   = { workspace = true }
futures                      = { workspace = true }
rolldown_core                = { path = "../rolldown_core" }
rolldown_error               = { path = "../rolldown_error" }
rolldown_plugin              = { path = "../rolldown_plugin" }
rolldown_plugin_node_resolve = { path = "../rolldown_plugin_node_resolve" }
rolldown_tracing             = { path = "../rolldown_tracing" }
sugar_path                   = { workspace = true }
tokio                        = { workspace = true, features = ["full"] }

[dev_dependencies]
criterion                    = "0.4.0"
insta                        = { workspace = true }
rolldown_plugin_node_resolve = { path = "../rolldown_plugin_node_resolve" }
rolldown_test_utils          = { path = "../rolldown_test_utils" }
testing_macros               = { workspace = true }

[[bench]]
harness = false
name    = "threejs"
