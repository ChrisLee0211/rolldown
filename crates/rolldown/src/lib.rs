mod output_options;
mod rolldown_build;
pub use output_options::*;
pub use rolldown_build::*;
pub use rolldown_core as core;
pub use rolldown_error as error;
mod input_options;
pub use input_options::*;
