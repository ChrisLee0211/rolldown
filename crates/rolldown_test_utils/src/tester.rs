use std::{
  collections::HashSet,
  path::{Path, PathBuf},
  sync::Arc,
};

use futures::FutureExt;
use rolldown_core::BuildError;

use crate::test_config::TestConfig;

pub struct Tester {
  pub config: TestConfig,
  pub warnings: Vec<BuildError>,
}

impl Tester {
  pub fn from_config_path(filepath: &Path) -> Self {
    let test_config = TestConfig::from_config_path(filepath);
    Self {
      config: test_config,
      warnings: Default::default(),
    }
  }

  pub fn input_options(&self, cwd: PathBuf) -> rolldown_core::InputOptions {
    rolldown_core::InputOptions {
      input: self.config.input.input.clone(),
      cwd,
      treeshake: self.config.input.treeshake,
      is_external: {
        let external = self
          .config
          .input
          .external
          .clone()
          .into_iter()
          .collect::<HashSet<_>>();
        Arc::new(move |specifier, _importer, _| {
          let external = external.clone();
          futures::future::ready(Ok(external.contains(specifier))).boxed()
        })
      },
      ..Default::default()
    }
  }
}
