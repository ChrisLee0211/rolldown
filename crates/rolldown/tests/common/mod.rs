use std::{
  path::{Path, PathBuf},
  str::FromStr,
};

use rolldown::Bundler;
use rolldown::{Asset, BuildResult, ExportMode, ModuleFormat, OutputOptions};
use rolldown_test_utils::tester::Tester;

pub struct CompiledFixture {
  pub tester: Tester,
  pub bundler: Bundler,
  pub output: BuildResult<Vec<Asset>>,
  pub dir_name: String,
  pub fixture_path: PathBuf,
}

impl CompiledFixture {
  pub fn output_friendly_to_snapshot(self) -> String {
    if self.tester.config.expect_error {
      let mut errors = self.output.unwrap_err().into_vec();
      errors.sort();
      let snapshot = [format!("---------- ERRORS ----------")]
        .into_iter()
        .chain(errors.iter().map(|err| {
          format!(
            "{}: {}",
            err.kind.code(),
            err.kind.to_readable_string(&self.fixture_path)
          )
        }))
        .collect::<Vec<_>>()
        .join("\n");

      return snapshot;
    }

    let mut assets = self.output.as_ref().unwrap().iter().collect::<Vec<_>>();
    assets.sort_by_key(|c| &c.filename);
    assets
      .iter()
      .flat_map(|asset| {
        [
          format!("---------- {} ----------", asset.filename),
          asset.content.trim().to_string(),
        ]
      })
      .chain(if self.tester.warnings.lock().unwrap().is_empty() {
        vec![]
      } else {
        let mut warnings = self.tester.warnings.lock().unwrap();
        warnings.sort();
        vec![
          format!("---------- WARNINGS ----------"),
          warnings
            .iter()
            .map(|w| {
              format!(
                "{}: {}",
                w.kind.code(),
                w.kind.to_readable_string(&self.fixture_path)
              )
            })
            .collect::<Vec<_>>()
            .join("\n"),
        ]
      })
      .collect::<Vec<_>>()
      .join("\n")
  }
}

pub async fn compile_fixture(test_config_path: &Path) -> CompiledFixture {
  let fixture_path = test_config_path.parent().unwrap();

  let tester = Tester::from_config_path(test_config_path);

  let mut bundler = Bundler::new(tester.input_options(fixture_path.to_path_buf()));

  if fixture_path.join("dist").is_dir() {
    std::fs::remove_dir_all(fixture_path.join("dist")).unwrap();
  }

  let output = bundler
    .generate(OutputOptions {
      // dir: Some(fixture_path.join("dist").to_string_lossy().to_string()),
      format: ModuleFormat::from_str(&tester.config.output.format).unwrap(),
      export_mode: ExportMode::from_str(&tester.config.output.export_mode).unwrap(),
      ..Default::default()
    })
    .await;
  let fixture_name = fixture_path
    .file_name()
    .unwrap()
    .to_string_lossy()
    .to_string();

  CompiledFixture {
    tester,
    bundler,
    output,
    dir_name: fixture_name,
    fixture_path: fixture_path.to_path_buf(),
  }
}

pub fn run_test(test_config_path: &Path) {
  // compile the fixture folder
  let compiled_fx = tokio::runtime::Runtime::new()
    .unwrap()
    .block_on(crate::common::compile_fixture(test_config_path));

  // If the test config has an expected error, assert that the error matches
  if let Some(expected_error) = compiled_fx.tester.config.expected_error {
    let errors = compiled_fx
      .output
      .expect_err("Expected error but got success")
      .into_vec();

    // TODO: should supports multiple errors
    let error = errors
      .iter()
      .find(|error| {
        error.kind.code() == expected_error.code
          && error.kind.to_readable_string(&compiled_fx.fixture_path) == expected_error.message
      })
      .unwrap_or_else(|| {
        eprintln!("errors: {errors:#?}");
        panic!("should find the error");
      });
    assert_eq!(error.kind.code(), expected_error.code);
    assert_eq!(
      error.kind.to_readable_string(&compiled_fx.fixture_path),
      expected_error.message
    );
    return;
  }

  // Otherwise, assert that the output matches the snapshot

  // Configure insta to use the test config path as the snapshot path
  let fixture_folder = test_config_path.parent().unwrap();
  let mut settings = insta::Settings::clone_current();
  settings.set_snapshot_path(fixture_folder);
  settings.set_prepend_module_to_snapshot(false);
  settings.set_input_file(fixture_folder);
  settings.bind(|| {
    insta::assert_snapshot!("output", compiled_fx.output_friendly_to_snapshot());
  });
}
