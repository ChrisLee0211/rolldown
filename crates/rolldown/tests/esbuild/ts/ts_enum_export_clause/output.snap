---
source: crates/rolldown/tests/common/mod.rs
expression: compiled_fx.output_friendly_to_snapshot()
input_file: crates/rolldown/tests/esbuild/ts/ts_enum_export_clause
---
---------- main.js ----------
// enums.ts
var A;
(function(A$1) {
    A$1[A$1["A"] = 1] = "A";
})(A || (A = {}));
var B;
(function(B$1) {
    B$1[B$1["B"] = 2] = "B";
})(B || (B = {}));
var C;
(function(C$1) {
    C$1[C$1["C"] = 3] = "C";
})(C || (C = {}));
var D;
(function(D$1) {
    D$1[D$1["D"] = 4] = "D";
})(D || (D = {}));

// main.ts
console.log([
    A.A,
    B.B,
    C.C,
    D.D
]);
